src/upkeeps/UpKeepManager.sol

```diff
@@ -0,0 +1,496 @@
+    /// @param _gasLimit amount of gas to provide the target contract when performing upkeep
+    /// @return linkAmount_ amount of LINK needed to cover the job
+    function _getLinkAmount(uint256 _gasLimit) internal view returns (uint256 linkAmount_) {
+        (, IKeeperRegistry.Config memory _c,) = CL_REGISTRY.getState();
+        (uint256 fastGasWei, uint256 linkEth) = _getFeedData();
+
+        uint256 adjustedGas = fastGasWei * _c.gasCeilingMultiplier;
+        uint256 weiForGas = adjustedGas * (_gasLimit + REGISTRY_GAS_OVERHEAD);
+        uint256 premium = PPB_BASE + _c.paymentPremiumPPB;
+
+        /// @dev amount of LINK to carry one `performUpKeep` operation
+        // See: _calculatePaymentAmount
+        // https://etherscan.io/address/0x02777053d6764996e594c3E88AF1D58D5363a2e6#code#F1#L776
+        linkAmount_ =
+        // From Wei to Eth * Premium / Ratio
+         ((weiForGas * (1e9) * (premium)) / (linkEth)) + (uint256(_c.flatFeeMicroLink) * (1e12));
+    }
+
+    /// @dev decodes the `bytes` calldata given by keepers and checks its validate against storage
+    /// @param _performData ABI-encoded member address to validate
+    function _validatePerformData(bytes calldata _performData) internal view returns (uint256 upKeepId_) {
+        address member = abi.decode(_performData, (address));
+
+        if (member == address(this)) {
+            upKeepId_ = monitoringUpKeepId;
+        } else {
+            upKeepId_ = membersInfo[member].upKeepId;
+        }
+
+        if (upKeepId_ == 0) revert MemberNotRegisteredYet(member);
+
+        (,,,,,, uint64 maxValidBlocknumber,) = CL_REGISTRY.getUpkeep(upKeepId_);
+        // https://etherscan.io/address/0x02777053d6764996e594c3e88af1d58d5363a2e6#code#F1#L332
+        if (maxValidBlocknumber != UINT64_MAX) revert UpkeepCancelled(upKeepId_);
+    }
+
+    /// @dev checks if upKeepId is under-funded, helper in `checkUpKeep`
+    /// and `performUpKeep` methods
+    /// @param _upKeepId task id to verify is underfunded
+    function _isUpKeepIdUnderFunded(uint256 _upKeepId)
+        internal
+        view
+        returns (uint96 minUpKeepBal_, bool underFunded_)
+    {
+        /// @dev check onchain the min and current amounts to consider top-up
+        minUpKeepBal_ = CL_REGISTRY.getMinBalanceForUpkeep(_upKeepId);
+        (,,, uint96 currentUpKeepBal,,,,) = CL_REGISTRY.getUpkeep(_upKeepId);
+
+        if (currentUpKeepBal <= minUpKeepBal_ * minRoundsTopUp) {
+            underFunded_ = true;
+        }
+    }
+
+    /// @notice carries over the top-up action of an member upKeep
+    /// @param _upKeepId id to verify if it is underfunded
+    function _topupUpkeep(uint256 _upKeepId) internal {
+        (uint96 minUpKeepBal, bool underFunded) = _isUpKeepIdUnderFunded(_upKeepId);
+
+        if (!underFunded) revert NotUnderFundedUpkeep(_upKeepId);
+
+        uint96 topupAmount = minUpKeepBal * uint96(roundsTopUp);
+
+        uint256 linkRegistryBal = LINK.balanceOf(address(this));
```

Comment:
Fine to leave as is, but worth just sending LINK to the contract so we can buy it via Cowswap

Also will save LINK fees